### PR TITLE
[FLOC-1300] Release Flocker 0.3.2+doc1

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1300

Pre-tag Review Checklist
- [ ] Do the acceptance tests. (although perhaps this can be skipped, as no changes were made to the running code)
- [ ] Check documentation.
  - [ ] The documentation should be available at https://docs.staging.clusterhq.com/en/0.3.2/
  - [ ] The following URLs should redirect to the above URL.
     - https://docs.staging.clusterhq.com/
     - https://docs.staging.clusterhq.com/en/
     - https://docs.staging.clusterhq.com/en/latest/
 
    In addition, check that deep-links to /en/latest/ work. https://docs.staging.clusterhq.com/en/latest/authors.html should redirect to https://docs.staging.clusterhq.com/en/0.3.2/authors.html
- [ ] Additional checks, since this is the first release on s3:
   - Check that, for all versions on [readthedocs](https://readthedocs.org/projects/flocker/), we redirect to the latest version, both for the root, and for deep links.

I'm not sure the above list is exhaustive.

The branch should not be merged until the release process is complete.

### CODE REVIEW

**Comment by wallrj** [comment]
The issue says "This will have the new theme, but not the new getting started guide" but I thought this  https://docs.staging.clusterhq.com/en/0.3.2/gettingstarted/index.html was the new getting started guide?

**Comment by tomprince** [comment]
Yeah. That should be changed I guess. The new getting started guide landed before this release happened. When the release was first started, that wasn't the case.

File: docs/_themes/clusterhq/layout_homepage.html [hunk 1]
```diff
@@ -117,8 +117,8 @@ <h2 class="heading text-center doc-banner">See our <a href="https://docs.cluster
 				<div><span class="widget-title">Documentation</span>			
 				<div class="textwidget">
 				<ul>
-					<li><a href="#">Getting Started</a></li>
-					<li><a href="https://docs.clusterhq.com/en/latest/gettinginvolved/index.html">Collaborate</a></li>
+					<li><a href="{{ pathto('gettingstarted/index') }}">Getting Started</a></li>
+					<li><a href="{{ pathto('gettinginvolved/contributing') }}">Collaborate</a></li>
```
Comment (wallrj) [inline docs/_themes/clusterhq/layout_homepage.html]: Looks like the "Collaborate" target  changed, but I agree that the contributing page is a better target.

**Comment by wallrj** [comment]
The 301 redirect `Location` headers have root relative URLs. I always thought they had to be absolute URLs....
```
(1250)(custom-404-FLOC-1250 ? =)[~/.../flocker/docs]$ curl -v https://docs.staging.clusterhq.com/
...
< HTTP/1.1 301 Moved Permanently
< Location: /en/latest/
```

...and are we sure about using permanent 301 redirect here. I'd have thought a temporary redirect would be more appropriate. And that's what RTD do.

```
(1250)(custom-404-FLOC-1250 ? =)[~/.../flocker/docs]$ curl -v https://flocker.readthedocs.org
...
< HTTP/1.1 302 FOUND
< Location: https://flocker.readthedocs.org/en/latest/
```

...and they use an absolute Location header too.

**Comment by tomprince** [comment]
https://en.wikipedia.org/wiki/HTTP_location and HTTPbis suggest that relative URIs are allowed, and work in practice.

**Comment by wallrj** [comment]
Oh wait, so 
 * `/` > `/en/latest/` (301), then
 * `/en/latest/` > `http://clusterhq-staging-docs.s3-website-us-east-1.amazonaws.com/en/0.3.2/` (302)

**Comment by tomprince** [comment]
Huh. That later one is wrong. :frowning:

**Comment by wallrj** [comment]
Also note that https://docs.staging.clusterhq.com/en/latest gives a 404.

**Comment by wallrj** [comment]
The versions that are currently listed on RTD all redirect to the latest version on our new docs website....but there may be links  to (now) unlisted versions of the documentation which still exist on RTD eg
 * http://flocker.readthedocs.org/en/0.1.0/
 * http://flocker.readthedocs.org/en/0.1.0/authors.html

**Comment by wallrj** [comment]
There's an un-encoded ampersand in the template
 * https://github.com/ClusterHQ/flocker/blob/release/flocker-0.3.2/docs/_themes/clusterhq/layout_docs.html#L14

**Comment by wallrj** [comment]
Thanks @tomprince 

I haven't had time to run the acceptance tests, but a few of the redirect URLs in your checklist didn't behave as expected.

I'll review again tomorrow unless someone gets there first.

**Comment by tomprince** [comment]
> There's an un-encoded ampersand in the template 

It looks like, according to the spec, since it is [unambiguous](https://html.spec.whatwg.org/multipage/syntax.html#syntax-ambiguous-ampersand), that it is valid. If you think this should still be fixed, please file an issue.

**Comment by tomprince** [comment]
> The versions that are currently listed on RTD all redirect to the latest version on our new docs website....but there may be links to (now) unlisted versions of the documentation which still exist on RTD

Is there a list of them? I guess maybe the ones with hashes [here](https://readthedocs.org/projects/flocker/versions/) (when logged in).

**Comment by tomprince** [comment]
I've update the list of redirects with that list.

**Comment by itamarst** [comment]
1. The redirects don't work from https://docs.staging.clusterhq.com/ and https://docs.staging.clusterhq.com/en/latest/.
2. Buildbot is red.
3. I don't really understand the RTD requirement.

**Comment by itamarst** [comment]
I guess buildbot being red isn't *necessarily* an issue? Since we're not building packages, just documentation? But if we're not building vagrant image this seems like it would fail.

**Comment by itamarst** [comment]
After discussion on Zulip we're building new vagrant image on a different buildbot.

Other notes:
1. Non-HTTPS URLs don't work on staging site. If this is true for production site that's a problem.
2. I tried checking redirects via another host, still seeing wrong values. Tom tells me it's fixed but still cached? But can't prove it's fixed.

**Comment by itamarst** [comment]
Our old tutorial URLs will be broken by this change; do we have redirects to deal with it?

**Comment by itamarst** [comment]
1. I verified manually old mongodb volume move tutorial still works manually (acceptance tests aren't available in this release, and master doesn't work with 0.3.2). Not really surprising, but it works.
2. Quickstart also works... modulo some blockers I filed.

**Comment by tomprince** [comment]
I've update the staging docs with update files, and invalidated cloudflare.

**Comment by wallrj** [comment]
I'll take another look and try the tutorial.

**Comment by wallrj** [comment]
Not sure it matters, but https://docs.staging.clusterhq.com/en/0.2.1/ doesn't redirect, instead returns a 0 byte response.

```
(1250)(release/flocker-0.3.2 ? =)[~/.../flocker/docs]$ curl -v https://docs.staging.clusterhq.com/en/0.2.1/ >/dev/null
* Adding handle: conn: 0x75fa10
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x75fa10) send_pipe: 1, recv_pipe: 0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* About to connect() to docs.staging.clusterhq.com port 443 (#0)
*   Trying 54.192.1.36...
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to docs.staging.clusterhq.com (54.192.1.36) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSL connection using TLS_RSA_WITH_AES_128_GCM_SHA256
* Server certificate:
* 	subject: CN=docs.staging.clusterhq.com,OU=Gandi Standard SSL,OU=Domain Control Validated
* 	start date: Jan 22 00:00:00 2015 GMT
* 	expire date: Jan 22 23:59:59 2016 GMT
* 	common name: docs.staging.clusterhq.com
* 	issuer: CN=Gandi Standard SSL CA 2,O=Gandi,L=Paris,ST=Paris,C=FR
> GET /en/0.2.1/ HTTP/1.1
> User-Agent: curl/7.32.0
> Host: docs.staging.clusterhq.com
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/octet-stream
< Content-Length: 0
< Connection: keep-alive
< Date: Wed, 04 Feb 2015 10:20:09 GMT
< Content-Language: en
< Last-Modified: Tue, 03 Feb 2015 23:08:03 GMT
< ETag: "d41d8cd98f00b204e9800998ecf8427e"
* Server AmazonS3 is not blacklisted
< Server: AmazonS3
< X-Cache: Miss from cloudfront
< Via: 1.1 02e93a1b0477fd0e9d0657e03ad4ab16.cloudfront.net (CloudFront)
< X-Amz-Cf-Id: 6kfhHp1N9ILEPaE2-sP6xH2unXJ_CXaUyrpG8cxGUZXo7cw5WUXlgw==
< 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host docs.staging.clusterhq.com left intact
```

**Comment by wallrj** [comment]
Some of the other versions list at https://readthedocs.org/projects/flocker/versions/ have the same problem:

```
(1250)(release/flocker-0.3.2 ? =)[~/.../flocker/docs]$ for version in 0.1.2 0.2.1 0.3.0  0.3.1 0.3.2; do url="https://docs.staging.clusterhq.com/en/${version}/"; echo $url; curl -o /dev/null  --silent -v "${url}" 2>&1 | grep '< HTTP/1\.1'; done
https://docs.staging.clusterhq.com/en/0.1.2/
< HTTP/1.1 200 OK
https://docs.staging.clusterhq.com/en/0.2.1/
< HTTP/1.1 200 OK
https://docs.staging.clusterhq.com/en/0.3.0/
< HTTP/1.1 301 Moved Permanently
https://docs.staging.clusterhq.com/en/0.3.1/
< HTTP/1.1 301 Moved Permanently
https://docs.staging.clusterhq.com/en/0.3.2/
< HTTP/1.1 200 OK
```

**Comment by wallrj** [comment]
It's really difficult to copy the command for pasting into the console. See attached.
screenshot from 2015-02-04 10 30 25

**Comment by wallrj** [comment]
Non-ssl links are not being redirected. Eg `flocker-deploy --help`  includes a non-ssl talk-to-us link. 
```
(flocker-tutorial)[vagrant@localhost ~]$ flocker-deploy --help
Usage: flocker-deploy [OPTIONS] DEPLOYMENT_CONFIGURATION_PATH APPLICATION_CONFIGURATION_PATH
If you have any issues or feedback, you can talk to us: http://docs.clusterhq.com/en/latest/gettinginvolved/contributing.html#talk-to-us
Options:
      --version  Print the program's version and exit.
      --help     Display this help and exit.
  -v, --verbose  Turn on verbose logging.

flocker-deploy allows you to configure existing nodes.

```

When I visit that on staging, http://docs.staging.clusterhq.com/en/latest/gettinginvolved/contributing.html#talk-to-us I get

```
ERROR
The request could not be satisfied.
Bad request.

Generated by cloudfront (CloudFront)
Request ID: 78ZK9WR8sz7mXYOk74H_enu6vK4yWWkrgN2eK0cCXeyacZLG6FhLSg==

```

**Comment by wallrj** [comment]
Even the stable (staging) docs https://docs.staging.clusterhq.com/en/0.3.2/ show the "You are reading an in-development version of the documentation" admonition.

Perhaps that's just because of the staging hostname?

**Comment by wallrj** [comment]
The vagrant node install instructions *assume* that you've got ``git`` installed and then spit a confusing error, because of the chain of commands

```

(flocker-tutorial)[vagrant@localhost ~]$ git clone \
>   https://github.com/clusterhq/vagrant-flocker && \
>   cd vagrant-flocker && \
>   vagrant up && \
>   [ -e "${SSH_AUTH_SOCK}" ] || eval $(ssh-agent) && \
>   ssh-add ~/.vagrant.d/insecure_private_key
-bash: git: command not found
Agent pid 1726
/home/vagrant/.vagrant.d/insecure_private_key: No such file or directory
```

**Comment by wallrj** [comment]
The 0.3.2 version of the AWS docs still point to the outdated get-clouds Fedora page
 * https://docs.staging.clusterhq.com/en/0.3.2/indepth/installation.html#aws-install
 * https://fedoraproject.org/en/get-fedora#clouds

**Comment by wallrj** [comment]
I followed the vagrant node instructions, but don't seem to have the redis or flask images pre-cached....which makes flocker-deploy slow....but let me double check that I'm getting the correct vagrant image.

```
[vagrant@node2 ~]$ sudo docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
busybox             latest              e72ac664f4f0        4 months ago        2.433 MB
clusterhq/mongodb   latest              1fe01bb7e1da        5 months ago        51.33 MB

```

**Comment by wallrj** [comment]
I found the Vagrant image built by @tomprince 's staged (old) buildbot.
 * http://build.staging.clusterhq.com/builders/flocker-vagrant-tutorial-box/builds/29

and verified that it contains the flask and redis images

```
(master ✕? =)[~/vagrant-flocker]$ vagrant ssh node1
Last login: Fri Sep 19 10:58:58 2014 from 10.0.2.2
Welcome to your Packer-built virtual machine.
[vagrant@node1 ~]$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[vagrant@node1 ~]$ sudo docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
dockerfile/redis    latest              c059fe93c46a        6 days ago          423.5 MB
clusterhq/flask     latest              fb9b1201592f        7 days ago          896.8 MB
busybox             latest              4986bf8c1536        4 weeks ago         2.433 MB
clusterhq/mongodb   latest              1fe01bb7e1da        5 months ago        51.33 MB
```

I suppose we could upload that box to Vagrant Cloud as a 0.3.2+doc1 image and make that the requirement in https://github.com/ClusterHQ/vagrant-flocker/blob/master/Vagrantfile

That would avoid making the tutorial image unnecessarily larger for the old tutorial.

**Comment by wallrj** [comment]
Yeah, using that version of the vagrant tutorial  image makes the flocker-deploy step much faster.

```
(flocker-tutorial)vagrant@vagrant-ubuntu-trusty-64:~/flocker-quickstart$ time flocker-deploy deployment-node1.yml fig.yml

real	0m3.745s
user	0m0.355s
sys	0m0.144s

```

**Comment by wallrj** [comment]
Shouldn't there be a final instruction to refresh your web browser and see that the application has been accessed 2 times....isn't that the point? The data hasn't been lost.
screenshot from 2015-02-04 12 42 40

**Comment by wallrj** [comment]
@jongiddy Noted that the syntax highlighting doesn't work for the wrapped command lines in the new gettingstarted page.
screenshot from 2015-02-04 15 37 45

**Pull request merged**